lib/Builtins.cpp

```diff
@@ -14,171 +14,339 @@
 
 #include "Builtins.h"
 
+#include <unordered_map>
+#include <stdlib.h>
+
 using namespace llvm;
+using namespace clspv;
+
+////////////////////////////////////////////////////////////////////////////////
+// utilities for const char* map keys
+struct cstr_hash {
+  std::size_t operator()(const char *cstr) const {
+    std::size_t hash = 5381;
+    for ( ; *cstr != '\0' ; ++cstr)
+      hash = (hash * 33) + *cstr;
+    return hash;
+  }
+};
+struct cstr_eq {
+  bool operator()(const char* a, const char *b) const { return strcmp(a, b) == 0; }
```

Comment:
Add a method comment:  // Returns true if the two C-style strings are equal.  Assumes they are both null-terminated.